### PR TITLE
MAINT: Switch to sparse arrays

### LINKED ISSUE #12642: MAINT: Switch to *_array instead of *_matrix for sparse
According to the SciPy folks it's more future compatible and possibly faster/better.

First step is probably to add support to h5io: https://github.com/h5io/h5io/pull/98.

Looks like the classes were added in SciPy 1.8 and our min is 1.9 so we should be good there.

### PR DESCRIPTION
**Pull request by larsoner**
Closes #12642

Eventually `*_matrix` will be deprecated, let's try to use `*_array` instead. TBD how much we have to deprecate vs just change directly.

### CODE REVIEW

**Comment by larsoner** [comment]
@drammock this one is ready I think!

File: doc/changes/devel/12646.bugfix.rst
Comment (drammock) [inline doc/changes/devel/12646.bugfix.rst:1]: This should go in the "notable changes" section (or whatever it is we call it)

File: mne/forward/forward.py [hunk 1]
```diff
@@ -838,11 +838,11 @@ def convert_forward_solution(
             fwd["surf_ori"] = True
         else:
             surf_rot = _block_diag(fwd["source_nn"].T, 3)
-            fwd["sol"]["data"] = fwd["_orig_sol"] * surf_rot
+            fwd["sol"]["data"] = fwd["_orig_sol"] @ surf_rot
             fwd["sol"]["ncol"] = 3 * fwd["nsource"]
             if fwd["sol_grad"] is not None:
                 x = sparse.block_diag([surf_rot] * 3)
-                fwd["sol_grad"]["data"] = fwd["_orig_sol_grad"] * x  # dot prod
+                fwd["sol_grad"]["data"] = fwd["_orig_sol_grad"] @ x  # dot prod
```
Comment (drammock) [inline mne/forward/forward.py:845]: Don't really need the line comment anymore, but ok to leave in too

File: mne/io/eeglab/tests/test_eeglab.py [hunk 2]
```diff
@@ -747,5 +747,5 @@ def test_eeglab_drop_nan_annotations(tmp_path):
         ch_types=np.repeat("EEG", len(ch_names)),
     )
 
-    with pytest.raises(RuntimeWarning, match="1 .* have an onset that is NaN.*"):
+    with pytest.warns(RuntimeWarning, match="1 .* have an onset that is NaN.*"):
```
Comment (drammock) [inline mne/io/eeglab/tests/test_eeglab.py:750]: Is this related?

File: mne/source_estimate.py [hunk 3]
```diff
@@ -3470,7 +3470,7 @@ def _prepare_label_extraction(stc, labels, src, mode, allow_empty, use_sparse):
             # Efficiency shortcut: use linearity early to avoid redundant
             # calculations
             elif mode == "mean":
-                vertidx = sparse.csr_matrix(vertidx.mean(axis=0))
+                vertidx = sparse.csr_array(vertidx.mean(axis=0)[np.newaxis])
```
Comment (drammock) [inline mne/source_estimate.py:3473]: (nitpick) Is `keepdims` an option here (instead of newaxis)? That feels cleaner to me for some reason... But not sure if our min version includes it

**Review by drammock — COMMENTED** [review]
Nice! In the end the diff was not that bad ( big but easy to review)

File: mne/io/eeglab/tests/test_eeglab.py (hunk 2 above)
Comment (larsoner) [reply mne/io/eeglab/tests/test_eeglab.py:750]: Related only because I made the `old` tests run more tests while I was doing some refactoring to make it easy for me to make the `old` env locally to debug a failure, and doing that exposed this issue. This always should have been a `pytest.warns`

File: mne/source_estimate.py (hunk 3 above)
Comment (larsoner) [reply mne/source_estimate.py:3473]: Indeed my first move was to try `keepdims` but it didn't work for some reason (maybe not on old SciPy or whatever?)
